Purchases/SwiftSources/LocalReceiptParsing/ReceiptParser.swift

```diff
@@ -0,0 +1,56 @@
+//
+//  ReceiptParser.swift
+//  Purchases
+//
+//  Created by Andrés Boedo on 7/22/20.
+//  Copyright © 2020 Purchases. All rights reserved.
+//
+
+import Foundation
+
+struct ReceiptParser {
+    private let objectIdentifierParser: ASN1ObjectIdentifierBuilder
+    private let containerBuilder: ASN1ContainerBuilder
+    private let receiptBuilder: AppleReceiptBuilder
+
+    init() {
+        self.objectIdentifierParser = ASN1ObjectIdentifierBuilder()
+        self.containerBuilder = ASN1ContainerBuilder()
+        self.receiptBuilder = AppleReceiptBuilder()
+    }
+
+    func parse(from data: Data) throws -> AppleReceipt {
+        let intData = [UInt8](data)
+
+        let asn1Container = try containerBuilder.build(fromPayload: ArraySlice(intData))
+        guard let receiptASN1Container = try findASN1Container(withObjectId: .data, inContainer: asn1Container) else {
+            throw ReceiptReadingError.dataObjectIdentifierMissing
+        }
+        let receipt = try receiptBuilder.build(fromASN1Container: receiptASN1Container)
+        return receipt
+    }
+}
+
+private extension ReceiptParser {
+    func findASN1Container(withObjectId objectId: ASN1ObjectIdentifier,
+                           inContainer container: ASN1Container) throws -> ASN1Container? {
+        if container.encodingType == .constructed {
+            var currentPayload = container.internalPayload
+            for internalContainer in container.internalContainers {
+                currentPayload = currentPayload.dropFirst(internalContainer.totalBytes)
```

Comment:
I am a little bit confused about this `dropFirst`. Why is it needed?

Also, you're only using `currentPayload ` in line 44. The next iteration, it's applying the `dropFirst` on the `currentPayload` created in the previous iteration? Why?

Reply:
per our convo the other day, I'm going to refactor this to remove the `currentPayload` stuff since it's no longer needed - it was needed before I extracted this method.